core/chains/evm/logpoller/log_poller.go

```diff
@@ -305,8 +305,28 @@ func (lp *logPoller) Filter(from, to *big.Int, bh *common.Hash) ethereum.FilterQ
 // Replay signals that the poller should resume from a new block.
 // Blocks until the replay is complete.
 // Replay can be used to ensure that filter modification has been applied for all blocks from "fromBlock" up to latest.
-func (lp *logPoller) Replay(ctx context.Context, fromBlock int64) error {
-	latest, err := lp.ec.HeadByNumber(ctx, nil)
+func (lp *logPoller) Replay(ctx context.Context, fromBlock int64, async bool) error {
+	replayCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	go func() {
+		// merge log poller context and parent context.  If either of them are cancelled, abort replay
+		// This is so that lp.run() can pass a single context to lp.PollAndSaveLogs().  Not used in this
+		// function since we want to return differentiate between the two, for returning proper error code
+		select {
+		case <-lp.ctx.Done():
+			cancel()
+		case <-ctx.Done():
+			if !async {
+				// In the async case, we don't want to cancel the replay after we've returned to the caller (unless the
+				// log poller is shutting down).  This is important for a web controller, since it will always
+				// cancel the context as soon as the request is complete, which will likely be before the replay is done.
+				cancel()
+			}
+		case <-replayCtx.Done():
+		}
+	}()
```

Comment:
This is extremely difficult to read and seems brittle since it tightly couples implementation details of systems that shouldn't know anything about each other. Why can't the caller own the context lifetime like usual?

Actually, why not simply this?
```suggestion
	if async {
		ctx = lp.ctx
	}
```

Reply:
The contexts need to be merged for the synchronous case, but you have a good point... I think for the async case it could be something similar to `ctx = lp.ctx`, although that would have to happen later not here.  Thanks for the suggestion, it's at least in the right direction.

This go routine for merging the contexts doesn't necessarily have to be spawned by this function, it could be moved into `lp.run()`.  I considered both possibilities and decided this one made slightly more sense, but I'm open to moving it if there's a good reason.

I've tried several simpler ideas to try to avoid merging the contexts, in earlier PR's, kept thinking there must be a better way.  Each of them ended up falling short in one way or another, so I closed the PR's.  I eventually decided that @connorwstein was right, that without a major re-arch of how the replays work in logpoller, this is the best option that does everything we need.  Let's discuss offline if you have other ideas, as it's probably too complex for github.

Reply:
> as it's probably too complex for github.

I disagree. This requires a nuanced discussion alongside the code.

Storing a context on the log poller in the first place is a misuse of the context API, but regardless of which type you store there is no need to merge the signals. The stored context only needs to be wired to things that don't already have one of their own - i.e. background routines literally using `context.Background()`. Incoming requests with a `context` arg will already be cancelled by the system on shutdown, or will encounter systems that have shut down and return an appropriate error. What difference does it make if we race to cancel the request context from another place?

Reply:
https://github.com/golang/go/issues/36503#issuecomment-642164302

> I guess I'm a little confused about why the request context wouldn't already have the server context as a parent to begin with.

https://github.com/golang/go/issues/36503#issuecomment-645573652

> Within Google's codebase, where the context package originated, we follow the rule that a context.Context should only be passed around via the call stack.
> ...
> This rule means that at any point in the call stack, there should be exactly one applicable Context, received as a function parameter. When following this pattern, the merge operation never makes sense.
>
> While merging context cancellation signals is straightforward, merging context values is not. Contexts can contain trace IDs and other information; which value would we pick when merging two contexts?